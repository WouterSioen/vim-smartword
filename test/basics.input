runtime! plugin/smartword.vim

function! CheckMapModes(lhs)
  let _ = ['n', 'x', 's', 'o', 'i', 'c', 'l']
  call map(_, 'maparg(' . string(a:lhs) . ', v:val)')
  call map(_, 'v:val != ""')
  return _
endfunction

function! CheckMovements(direction, movement_command)
  let cols = []

  tabnew
    call setline(1, ['', '<a href="http://www.vim.org/">www.vim.org</a>', ''])

    if a:direction ==# 'forward'
      normal! 2gg0
    else
      normal! 2gg$
    endif
    while line('.') == 2
      call add(cols, col('.'))
      silent! execute a:movement_command
    endwhile
  tabclose!

  return cols
endfunction

function! CheckOperations(direction, operation_command)
  let cols = []

  tabnew
    call setline(1, ['', '<a href="http://www.vim.org/">www.vim.org</a>', ''])

    if a:direction ==# 'forward'
      normal! 2gg0
    else
      normal! 2gg$
    endif
    let last_col = -1
    while line('.') == 2 && col('.') != last_col
      let last_col = col('.')
      call add(cols, last_col)
      silent! execute a:operation_command
    endwhile
  tabclose!

  return cols
endfunction

function s:describe__UI_key_mapping_definitions()  "{{{1
  It should be defined properly.

  Should CheckMapModes('<Plug>(smartword-w)') ==# [1, 1, 1, 1, 0, 0, 0]
  Should CheckMapModes('<Plug>(smartword-b)') ==# [1, 1, 1, 1, 0, 0, 0]
  Should CheckMapModes('<Plug>(smartword-e)') ==# [1, 1, 1, 1, 0, 0, 0]
  Should CheckMapModes('<Plug>(smartword-ge)') ==# [1, 1, 1, 1, 0, 0, 0]
endfunction




function s:describe__UI_key_mappings_with_selection_inclusive()  "{{{1
  set selection=inclusive

  It should move the cursor properly in Normal mode.

  Should CheckMovements('forward', 'normal! w') == [
  \   1, 2, 4, 8, 10,
  \   14, 17, 20, 21, 24,
  \   25, 28, 31, 34, 35,
  \   38, 39, 42, 44, 45
  \ ]
  Should CheckMovements('forward', "normal \<Plug>(smartword-w)") == [
  \   1, 2, 4, 10, 17,
  \   21, 25, 31, 35, 39,
  \   44
  \ ]

  Should CheckMovements('backward', 'normal! b') == [
  \   45, 44, 42, 39, 38,
  \   35, 34, 31, 28, 25,
  \   24, 21, 20, 17, 14,
  \   10, 8, 4, 2, 1
  \ ]
  Should CheckMovements('backward', "normal \<Plug>(smartword-b)") == [
  \   45, 44, 39, 35, 31,
  \   25, 21, 17, 10, 4,
  \   2
  \ ]

  Should CheckMovements('forward', 'normal! e') == [
  \   1, 2, 7, 9, 13,
  \   16, 19, 20, 23, 24,
  \   27, 30, 33, 34, 37,
  \   38, 41, 43, 44, 45
  \ ]
  Should CheckMovements('forward', "normal \<Plug>(smartword-e)") == [
  \   1, 2, 7, 13, 19,
  \   23, 27, 33, 37, 41,
  \   44
  \ ]

  Should CheckMovements('backward', 'normal! ge') == [
  \   45, 44, 43, 41, 38,
  \   37, 34, 33, 30, 27,
  \   24, 23, 20, 19, 16,
  \   13, 9, 7, 2, 1
  \ ]
  Should CheckMovements('backward', "normal \<Plug>(smartword-ge)") == [
  \   45, 44, 41, 37, 33,
  \   27, 23, 19, 13, 7,
  \   2
  \ ]

  It should move the cursor properly in Visual mode.

  Should CheckMovements('forward', 'normal! vwv') == [
  \   1, 2, 4, 8, 10,
  \   14, 17, 20, 21, 24,
  \   25, 28, 31, 34, 35,
  \   38, 39, 42, 44, 45
  \ ]
  Should CheckMovements('forward', "normal v\<Plug>(smartword-w)v") == [
  \   1, 2, 4, 10, 17,
  \   21, 25, 31, 35, 39,
  \   44
  \ ]

  Should CheckMovements('backward', 'normal! vbv') == [
  \   45, 44, 42, 39, 38,
  \   35, 34, 31, 28, 25,
  \   24, 21, 20, 17, 14,
  \   10, 8, 4, 2, 1
  \ ]
  Should CheckMovements('backward', "normal v\<Plug>(smartword-b)v") == [
  \   45, 44, 39, 35, 31,
  \   25, 21, 17, 10, 4,
  \   2
  \ ]

  Should CheckMovements('forward', 'normal! vev') == [
  \   1, 2, 7, 9, 13,
  \   16, 19, 20, 23, 24,
  \   27, 30, 33, 34, 37,
  \   38, 41, 43, 44, 45
  \ ]
  Should CheckMovements('forward', "normal v\<Plug>(smartword-e)v") == [
  \   1, 2, 7, 13, 19,
  \   23, 27, 33, 37, 41,
  \   44
  \ ]

  Should CheckMovements('backward', 'normal! vgev') == [
  \   45, 44, 43, 41, 38,
  \   37, 34, 33, 30, 27,
  \   24, 23, 20, 19, 16,
  \   13, 9, 7, 2, 1
  \ ]
  Should CheckMovements('backward', "normal v\<Plug>(smartword-ge)v") == [
  \   45, 44, 41, 37, 33,
  \   27, 23, 19, 13, 7,
  \   2
  \ ]

  It should move the cursor properly in Operator-pending mode.

  Should CheckOperations('forward', 'normal! yw`]') == [
  \   1, 2, 4, 8, 10,
  \   14, 17, 20, 21, 24,
  \   25, 28, 31, 34, 35,
  \   38, 39, 42, 44, 45
  \ ]
  Should CheckOperations('forward', "normal y\<Plug>(smartword-w)`]") == [
  \   1, 2, 4, 10, 17,
  \   21, 25, 31, 35, 39,
  \   44, 45
  \ ]

  Should CheckOperations('backward', 'normal! yb') == [
  \   45, 44, 42, 39, 38,
  \   35, 34, 31, 28, 25,
  \   24, 21, 20, 17, 14,
  \   10, 8, 4, 2, 1
  \ ]
  Should CheckOperations('backward', "normal y\<Plug>(smartword-b)") == [
  \   45, 44, 39, 35, 31,
  \   25, 21, 17, 10, 4,
  \   2
  \ ]

  Should CheckOperations('forward', 'normal! ye`]') == [
  \   1, 2, 7, 9, 13,
  \   16, 19, 20, 23, 24,
  \   27, 30, 33, 34, 37,
  \   38, 41, 43, 44, 45
  \ ]
  Should CheckOperations('forward', "normal y\<Plug>(smartword-e)`]") == [
  \   1, 2, 7, 13, 19,
  \   23, 27, 33, 37, 41,
  \   44
  \ ]

  Should CheckOperations('backward', 'normal! yge') == [
  \   45, 44, 43, 41, 38,
  \   37, 34, 33, 30, 27,
  \   24, 23, 20, 19, 16,
  \   13, 9, 7, 2, 1
  \ ]
  Should CheckOperations('backward', "normal y\<Plug>(smartword-ge)") == [
  \   45, 44, 41, 37, 33,
  \   27, 23, 19, 13, 7,
  \   2
  \ ]

  set selection&
endfunction




function s:describe__UI_key_mappings_with_selection_exclusive()  "{{{1
  set selection=exclusive

  It should move the cursor properly in Normal mode.

  Should CheckMovements('forward', 'normal! w') == [
  \   1, 2, 4, 8, 10,
  \   14, 17, 20, 21, 24,
  \   25, 28, 31, 34, 35,
  \   38, 39, 42, 44, 45
  \ ]
  Should CheckMovements('forward', "normal \<Plug>(smartword-w)") == [
  \   1, 2, 4, 10, 17,
  \   21, 25, 31, 35, 39,
  \   44
  \ ]

  Should CheckMovements('backward', 'normal! b') == [
  \   45, 44, 42, 39, 38,
  \   35, 34, 31, 28, 25,
  \   24, 21, 20, 17, 14,
  \   10, 8, 4, 2, 1
  \ ]
  Should CheckMovements('backward', "normal \<Plug>(smartword-b)") == [
  \   45, 44, 39, 35, 31,
  \   25, 21, 17, 10, 4,
  \   2
  \ ]

  Should CheckMovements('forward', 'normal! e') == [
  \   1, 2, 7, 9, 13,
  \   16, 19, 20, 23, 24,
  \   27, 30, 33, 34, 37,
  \   38, 41, 43, 44, 45
  \ ]
  Should CheckMovements('forward', "normal \<Plug>(smartword-e)") == [
  \   1, 2, 7, 13, 19,
  \   23, 27, 33, 37, 41,
  \   44
  \ ]

  Should CheckMovements('backward', 'normal! ge') == [
  \   45, 44, 43, 41, 38,
  \   37, 34, 33, 30, 27,
  \   24, 23, 20, 19, 16,
  \   13, 9, 7, 2, 1
  \ ]
  Should CheckMovements('backward', "normal \<Plug>(smartword-ge)") == [
  \   45, 44, 41, 37, 33,
  \   27, 23, 19, 13, 7,
  \   2
  \ ]

  It should move the cursor properly in Visual mode.

  Should CheckMovements('forward', 'normal! vwv') == [
  \   1, 2, 4, 8, 10,
  \   14, 17, 20, 21, 24,
  \   25, 28, 31, 34, 35,
  \   38, 39, 42, 44, 45
  \ ]
  Should CheckMovements('forward', "normal v\<Plug>(smartword-w)v") == [
  \   1, 2, 4, 10, 17,
  \   21, 25, 31, 35, 39,
  \   44
  \ ]

  Should CheckMovements('backward', 'normal! vbv') == [
  \   45, 44, 42, 39, 38,
  \   35, 34, 31, 28, 25,
  \   24, 21, 20, 17, 14,
  \   10, 8, 4, 2, 1
  \ ]
  Should CheckMovements('backward', "normal v\<Plug>(smartword-b)v") == [
  \   45, 44, 39, 35, 31,
  \   25, 21, 17, 10, 4,
  \   2
  \ ]

  Should CheckMovements('forward', 'normal! vehv') == [
  \   1, 2, 7, 9, 13,
  \   16, 19, 20, 23, 24,
  \   27, 30, 33, 34, 37,
  \   38, 41, 43, 44, 45
  \ ]
  Should CheckMovements('forward', "normal v\<Plug>(smartword-e)hv") == [
  \   1, 2, 7, 13, 19,
  \   23, 27, 33, 37, 41,
  \   44
  \ ]

  Should CheckMovements('backward', 'normal! vgev') == [
  \   45, 44, 43, 41, 38,
  \   37, 34, 33, 30, 27,
  \   24, 23, 20, 19, 16,
  \   13, 9, 7, 2, 1
  \ ]
  Should CheckMovements('backward', "normal v\<Plug>(smartword-ge)v") == [
  \   45, 44, 41, 37, 33,
  \   27, 23, 19, 13, 7,
  \   2
  \ ]

  It should move the cursor properly in Operator-pending mode.

  Should CheckOperations('forward', 'normal! yw`]') == [
  \   1, 2, 4, 8, 10,
  \   14, 17, 20, 21, 24,
  \   25, 28, 31, 34, 35,
  \   38, 39, 42, 44, 45
  \ ]
  Should CheckOperations('forward', "normal y\<Plug>(smartword-w)`]") == [
  \   1, 2, 4, 10, 17,
  \   21, 25, 31, 35, 39,
  \   44, 45
  \ ]

  Should CheckOperations('backward', 'normal! yb') == [
  \   45, 44, 42, 39, 38,
  \   35, 34, 31, 28, 25,
  \   24, 21, 20, 17, 14,
  \   10, 8, 4, 2, 1
  \ ]
  Should CheckOperations('backward', "normal y\<Plug>(smartword-b)") == [
  \   45, 44, 39, 35, 31,
  \   25, 21, 17, 10, 4,
  \   2
  \ ]

  Should CheckOperations('forward', 'normal! ye`]') == [
  \   1, 2, 7, 9, 13,
  \   16, 19, 20, 23, 24,
  \   27, 30, 33, 34, 37,
  \   38, 41, 43, 44, 45
  \ ]
  Should CheckOperations('forward', "normal y\<Plug>(smartword-e)`]") == [
  \   1, 2, 7, 13, 19,
  \   23, 27, 33, 37, 41,
  \   44
  \ ]

  Should CheckOperations('backward', 'normal! yge') == [
  \   45, 44, 43, 41, 38,
  \   37, 34, 33, 30, 27,
  \   24, 23, 20, 19, 16,
  \   13, 9, 7, 2, 1
  \ ]
  Should CheckOperations('backward', "normal y\<Plug>(smartword-ge)") == [
  \   45, 44, 41, 37, 33,
  \   27, 23, 19, 13, 7,
  \   2
  \ ]

  set selection&
endfunction




" __END__  "{{{1
" vim: filetype=vim foldmethod=marker
