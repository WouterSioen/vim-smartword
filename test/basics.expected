==== UI_key_mapping_definitions 
---- It should be defined properly. 
....

==== UI_key_mappings_with_selection_inclusive 
---- It should move the cursor properly in Normal mode. 
........
---- It should move the cursor properly in Visual mode. 
........
---- It should move the cursor properly in Operator-pending mode. 
........



**** Result ****


28 examples, 0 failures
